cachi2/core/package_managers/yarn/locators.py

```diff
@@ -1,115 +1,273 @@
+        patch = patch.removeprefix("~")  # '~' denotes an optional patch
+        if re.match(r"^builtin<([^>]+)>$", patch):
+            return patch
+        else:
+            return Path(patch)
+
+    patches = [process_patch_path(p) for p in reference.selector.split("&")]
+    if locator_param := _get_param(reference, "locator"):
+        parent_locator = parse_locator(locator_param)
+    else:
+        parent_locator = None
+
+    return PatchLocator(original_package, patches, parent_locator)
+
+
+def _parse_file_locator(locator: "_ParsedLocator") -> FileLocator:
+    reference = locator.parsed_reference
+
+    relpath = Path(reference.selector)
+    # for 'file:' directories, Yarnberry uses the path as both source and selector
+    # https://github.com/yarnpkg/berry/blob/b6026842dfec4b012571b5982bb74420c7682a73/packages/plugin-file/sources/fileUtils.ts#L16
+    if reference.source and Path(reference.source) != relpath:
+        raise UnexpectedFormat("conflicting paths in locator")
+
+    locator_param = _get_param(reference, "locator")
+    if not locator_param:
+        raise UnexpectedFormat("missing 'locator' param")
+
+    parent_locator = parse_locator(locator_param)
+    return FileLocator(relpath, parent_locator)
+
+
+def _get_param(reference: "_ParsedReference", param_name: str) -> Optional[str]:
+    if not reference.params or not (param_value := reference.params.get(param_name)):
+        return None
+    if len(param_value) != 1:
+        raise UnexpectedFormat(f"expected 1 {param_name!r} param, got {len(param_value)}")
+    return param_value[0]
+
+
+# --- Parsing locators generically ---
+
+
+# dataclass rather than NamedTuple because NamedTuple doesn't support cached_property
+@dataclass(frozen=True)
+class _ParsedLocator:
+    scope: Optional[str]
+    name: str
+    raw_reference: str
+
+    def __str__(self) -> str:
+        name_at_ref = f"{self.name}@{self.raw_reference}"
+        if self.scope:
+            return f"@{self.scope}/{name_at_ref}"
+        return name_at_ref
+
+    @cached_property
+    def parsed_reference(self) -> "_ParsedReference":
+        return _parse_reference(self.raw_reference)
+
+
+class _ParsedReference(NamedTuple):
+    protocol: Optional[str]
+    source: Optional[str]
+    selector: str
+    params: Optional[dict[str, list[str]]]
+
+
+def _parse_locator(locator_str: str) -> _ParsedLocator:
+    # https://github.com/yarnpkg/berry/blob/b6026842dfec4b012571b5982bb74420c7682a73/packages/yarnpkg-core/sources/structUtils.ts#L411
+    locator_re = re.compile(r"^(?:@([^/]+?)/)?([^@/]+?)(?:@(.+))$")
+    match = locator_re.match(locator_str)
+    if not match:
+        raise UnexpectedFormat("could not parse locator (expected [@scope/]name@reference)")
+    scope, name, reference = match.groups()
+    return _ParsedLocator(scope, name, reference)
+
+
+def _parse_reference(reference_str: str) -> _ParsedReference:
+    """Parse a reference string.
```

Comment:
Maybe add a small note on what a reference string is (the part after the @ in a locator?)

Reply:
added